langpuppet/gapi.go

```diff
@@ -0,0 +1,327 @@
+		Debug:         obj.data.Debug,
+		Logf: func(format string, v ...interface{}) {
+			obj.data.Logf(puppet.Name+": "+format, v...)
+		},
+	}
+
+	if err := obj.langGAPI.Init(dataLang); err != nil {
+		return err
+	}
+	if err := obj.puppetGAPI.Init(dataPuppet); err != nil {
+		return err
+	}
+
+	obj.closeChan = make(chan struct{})
+	obj.initialized = true
+	return nil
+}
+
+// Graph returns a current Graph.
+func (obj *GAPI) Graph() (*pgraph.Graph, error) {
+	if !obj.initialized {
+		return nil, fmt.Errorf("%s: GAPI is not initialized", Name)
+	}
+
+	var graphErr error
+
+	obj.graphFlagMutex.Lock()
+	if obj.langGraphReady {
+		obj.langGraphReady = false
+		obj.graphFlagMutex.Unlock()
+		obj.currentLangGraph, graphErr = obj.langGAPI.Graph()
+		if graphErr != nil {
+			return nil, graphErr
+		}
+	} else {
+		obj.graphFlagMutex.Unlock()
+	}
+
+	obj.graphFlagMutex.Lock()
+	if obj.puppetGraphReady {
+		obj.puppetGraphReady = false
+		obj.graphFlagMutex.Unlock()
+		obj.currentPuppetGraph, graphErr = obj.puppetGAPI.Graph()
+		if graphErr != nil {
+			return nil, graphErr
+		}
+	} else {
+		obj.graphFlagMutex.Unlock()
+	}
+
+	g, err := mergeGraphs(obj.currentLangGraph, obj.currentPuppetGraph)
+
+	if obj.data.Debug {
+		obj.currentLangGraph.Logf(func(format string, v ...interface{}) {
+			obj.data.Logf("graph: "+lang.Name+": "+format, v...)
+		})
+		obj.currentPuppetGraph.Logf(func(format string, v ...interface{}) {
+			obj.data.Logf("graph: "+puppet.Name+": "+format, v...)
+		})
+		if err == nil {
+			g.Logf(func(format string, v ...interface{}) {
+				obj.data.Logf("graph: "+Name+": "+format, v...)
+			})
+		}
+	}
+
+	return g, err
+}
+
+// Next returns nil errors every time there could be a new graph.
+func (obj *GAPI) Next() chan gapi.Next {
+	ch := make(chan gapi.Next)
+	obj.wg.Add(1)
+	go func() {
+		defer obj.wg.Done()
+		defer close(ch) // this will run before the obj.wg.Done()
+		if !obj.initialized {
+			next := gapi.Next{
+				Err:  fmt.Errorf("%s: GAPI is not initialized", Name),
+				Exit: true, // exit, b/c programming error?
+			}
+			ch <- next
+			return
+		}
+		nextLang := obj.langGAPI.Next()
+		nextPuppet := obj.puppetGAPI.Next()
+
+		firstLang := false
+		firstPuppet := false
+
+		for {
+			select {
+			case nextChild := <-nextLang:
+				if nextChild.Err != nil {
+					ch <- nextChild
+				}
+				if nextChild.Exit {
+					return
+				}
+				obj.graphFlagMutex.Lock()
+				obj.langGraphReady = true
+				obj.graphFlagMutex.Unlock()
+				firstLang = true
+			case nextChild := <-nextPuppet:
+				if nextChild.Err != nil {
+					ch <- nextChild
+				}
+				if nextChild.Exit {
+					return
+				}
+				obj.graphFlagMutex.Lock()
+				obj.puppetGraphReady = true
+				obj.graphFlagMutex.Unlock()
+				firstPuppet = true
+			case <-obj.closeChan:
+				return
+			}
+
+			if !firstLang || !firstPuppet {
+				continue
+			}
+
+			obj.data.Logf("generating new composite graph...")
+			next := gapi.Next{
+				Exit: false,
+				Err:  nil,
```

Comment:
err goes here :)
you might want to do this too:

https://github.com/purpleidea/mgmt/blob/28f343ac502971754a5be586e22f7c2656a17748/lang/gapi.go#L268